haystack/nodes/retriever/web.py

```diff
@@ -168,106 +188,132 @@ def retrieve(  # type: ignore[override]
+        self, query_norm: str, preprocessor: Optional[PreProcessor], top_k: Optional[int]
+    ) -> List[Document]:
+        """
+        Retrieve documents from the web based on the query.
+
+        :param query_norm: The normalized query string.
+        :param preprocessor: The PreProcessor to be used to split documents into paragraphs.
+        :param top_k: The number of documents to be returned by the retriever. If None, the default value is used.
+        :return: List of Document objects.
+        """
+
+        search_results, _ = self.web_search.run(query=query_norm)
+        search_results_docs = search_results["documents"]
+        if self.mode == "snippets":
+            return search_results_docs[:top_k]  # type: ignore
+
+        links: List[SearchResult] = self._prepare_links(search_results_docs)
+        logger.debug("Starting to fetch %d links from WebSearch results", len(links))
+        return self._scrape_links(links, query_norm, preprocessor)
+
+    def _prepare_links(self, search_results: List[Document]) -> List[SearchResult]:
+        """
+        Prepare a list of SearchResult objects based on the search results from the search engine.
 
-        logger.debug("Processed %d documents resulting in %s documents", len(extracted_docs), len(processed_docs))
-        return processed_docs[:top_k]
+        :param search_results: List of Document objects obtained from web search.
+        :return: List of SearchResult objects.
+        """
+        if not search_results:
+            return []
+
+        links: List[SearchResult] = [
+            SearchResult(r.meta["link"], r.content, r.meta.get("score"), r.meta.get("position"))
+            for r in search_results
+            if r.meta.get("link")
+        ]
+        return links
+
+    def _scrape_links(
+        self, links: List[SearchResult], query_norm: str, preprocessor: Optional[PreProcessor]
+    ) -> List[Document]:
+        """
+        Scrape the links and return the documents.
+
+        :param links: List of SearchResult objects.
+        :param query_norm: The normalized query string.
+        :param preprocessor: The PreProcessor object to be used to split documents into paragraphs.
+        :return: List of Document objects obtained from scraping the links.
+        """
+        if not links:
+            return []
+
+        def scrape_direct(link: SearchResult) -> List[Document]:
+            reader = (
+                LinkContentRetriever(pre_processor=preprocessor)
+                if self.mode == "preprocessed_documents" and preprocessor
+                else LinkContentRetriever()
+            )
+            return reader(
+                url=link.url,
+                doc_kwargs={
+                    "search.score": link.score,
+                    "id_hash_keys": ["meta.url", "meta.search.query"],
+                    "search.query": query_norm,
+                    "search.position": link.position,
+                    "text": link.snippet,
```

Comment:
To answer https://github.com/deepset-ai/haystack/pull/5227#discussion_r1261274381 @anakin87 it is mostly because of this line as here we pass the link snippet under "text" key. So we could use any key that makes sense. So if the LinkContentFetcher doesn't fetch the page - it would be useful to have some fallback. The question is how do we settle on the standard for this key? Here it is simply "text" but could be "content" or anything for that matter.